File: P121/P121.cpp
#include <iostream>

using namespace std;

class SolutionGreedy {
public:
	int maxProfit(vector<int>& prices) {
		
		const int size = prices.size();

		if (size < 2) {
			return 0;
		}
		
		int maxProfit = 0;
		
		int minPrice = prices[0];

		for (int i = 1; i < size; i++) {
			
			const int current = prices[i];
			
			if (current - minPrice > maxProfit) {
				maxProfit = current - minPrice;
			}
			
			if (current < minPrice) {
				minPrice = current;
			}
		}
		
		return maxProfit;
	}
};

int main(int argc, char *argv[]) {
	
}

File: P121/P121.java
class P121 {
	public static void main(String[] args) {
		
		int[] input = {4, 2, 3, 4, 5};
		int result = maxProfit(input);
		
		System.out.println(result);
	}

	public static int maxProfit(int[] prices) {
		
		int minprice = Integer.MAX_VALUE;
		int maxprofit = 0;
		
		for (int i = 0; i < prices.length; i++) {
			
			int profit = prices[i] - minprice;
			if (profit > maxprofit) {
				maxprofit = profit;
			}
			if (prices[i] < minprice) {
				minprice = prices[i];
			}
		}
		
		return maxprofit;
	}
}
